htmresearch/frameworks/sp_paper/sp_metrics.py

```diff
@@ -755,4 +779,142 @@ def calculateInputSpaceCoverage(sp):
+  return inputSpaceCoverage
+
+
+def reconstructionError(sp, inputVectors, activeColumnVectors, threshold=0.):
+  """
+  Computes a reconstruction error. The reconstuction $r(x)$ of an input vector $x$
+  is given by the sum of the active column's connected synapses vector of 
+  the SDR representation $sdr(x)$ of $x$ normalized by $1/numActiveColumns$. 
+  The error is the normalized sum over the "hamming distance" (i.e. distance 
+  induced by L1 norm) of $x$ and its reconstruction $r(x)$, i.e. (mathy stuff in LaTex)
+  \[
+      Reconstruction Error = (1/batchSize) * \sum_{x \in InputBatch} \| x - r(x) \|_1 .
+  \]
+  Note that $r(x)$ can be expressed as
+  \[
+      r(x) = (1/sdrSize) * C * sdr(x) ,
+  \]
+  where we view $sdr(x)$ as a binary column vector and $C$ is the 
+  binary matrix whose jth column encodes the synaptic connectivity of 
+  the pooler's columns and the input bits, i.e. 
+  \[
+        c_{i,j} = 1   :<=>  column j has a stable synaptic 
+                            connection to input bit i.
+  \]  
+  """
+  batchSize        = inputVectors.shape[0]
+  normalizingConst = batchSize
+  numColumns       = np.prod(sp.getColumnDimensions())
+  SDRSize          = sp.getLocalAreaDensity() * numColumns
+  connectionMatrix = getConnectedSyns(sp)
+
+  reconstructionVectors = np.dot(activeColumnVectors, connectionMatrix)
+  reconstructionVectors = reconstructionVectors/SDRSize
+  
+  if threshold > 0.:
+    reconstructionVectors =np.where( 
+                            reconstructionVectors > threshold, 
+                            np.ones( reconstructionVectors.shape), 
+                            np.zeros(reconstructionVectors.shape))
+
+  Err = np.sum(np.absolute(reconstructionVectors  - inputVectors))
+
+  return Err/normalizingConst
+
+
+def witnessError(sp, activeColumnsCurrentEpoch, inputVectors):
+  """
+  Computes a variation of a reconstruction error. It measures the average 
+  hamming distance of an active column's connected synapses vector and its witnesses. 
+  An input vector is called witness for a column, iff the column is among 
+  the active columns for the input computed by the spatial pooler. 
+  The error is given by
+  \[     
+      Witness Error = (1/batchSize) * \sum_{x \in InputBatch} 
+                          (1/sdrSize) * \sum_{i \in sdr(x)} \| x -  syn(i) \|_1.
+  \]
+  Note: Turns out that in our setting (x and syn(i) binary vectors) we have 
+  \[
+      Witness Error = Reconstruction Error.
+  \]
+  """
+  connectionMatrix = getConnectedSyns(sp)
+  batchSize        = inputVectors.shape[0]
+  normalizingConst = batchSize
+  # 1st sum... over each input in batch
+  Err = 0.
+  for i in range(batchSize):
+    activeColumns = np.where(activeColumnsCurrentEpoch[i] > 0.)[0]
+    numActive     = activeColumns.shape[0]
+    # 2nd sum... over each active colum
+    err = 0.
+    for j in activeColumns:
+      # Compute hamming distance and accumulate
+      connectionVector = connectionMatrix[j]
```

Comment:
No need to define connectionVector if it is used only once. It could be simplified as
`err += np.sum(np.absolute(connectionVector[j] - inputVectors[i]))`